### ISSUE TITLE
uses != reuses | k=8,m=3 EC pool

### ISSUE THREAD

**Issue by BonzTM** [TheJJ/ceph-balancer#10]
Hello!

I've been trying to use the balancer on my EC pool, and I am receiving the following error:
```sh
Traceback (most recent call last):
  File "./placementoptimizer.py", line 2080, in <module>
    try_pg_move.prepare_crush_check()
  File "./placementoptimizer.py", line 1060, in prepare_crush_check
    raise Exception(f"during emit, rule step {idx} item {item} was used {uses} != {reuses} expected")
Exception: during emit, rule step 0 item -15 was used 11 != 12 expected
```

I am assuming that during the sanity check, because I have 11 chunks of data (k8m3) but the crush rules specify 4hosts/3osds (12 chunks), this is why I am getting this error?

I am using Proxmox 7.2, CEPH Pacific 16.2.7.  Here is my **CRUSH map**.

```sh
# begin crush map
tunable choose_local_tries 0
tunable choose_local_fallback_tries 0
tunable choose_total_tries 50
tunable chooseleaf_descend_once 1
tunable chooseleaf_vary_r 1
tunable chooseleaf_stable 1
tunable straw_calc_version 1
tunable allowed_bucket_algs 54

# devices
device 0 osd.0 class nvme
device 1 osd.1 class nvme
device 2 osd.2 class nvme
device 3 osd.3 class nvme
device 4 osd.4 class hdd
device 5 osd.5 class hdd
device 6 osd.6 class hdd
device 7 osd.7 class hdd
device 8 osd.8 class hdd
device 9 osd.9 class hdd
device 10 osd.10 class hdd
device 11 osd.11 class hdd
device 12 osd.12 class hdd
device 13 osd.13 class hdd
device 14 osd.14 class hdd
device 15 osd.15 class hdd
device 16 osd.16 class hdd
device 17 osd.17 class hdd
device 18 osd.18 class hdd
device 19 osd.19 class hdd
device 20 osd.20 class hdd
device 21 osd.21 class hdd
device 22 osd.22 class hdd
device 23 osd.23 class hdd
device 24 osd.24 class hdd
device 25 osd.25 class hdd
device 26 osd.26 class hdd
device 27 osd.27 class hdd
device 28 osd.28 class hdd
device 29 osd.29 class hdd
device 30 osd.30 class hdd
device 31 osd.31 class hdd
device 32 osd.32 class nvme
device 33 osd.33 class nvme
device 34 osd.34 class nvme
device 35 osd.35 class nvme
device 36 osd.36 class hdd
device 37 osd.37 class hdd
device 38 osd.38 class hdd
device 39 osd.39 class hdd
device 40 osd.40 class nvme
device 41 osd.41 class nvme
device 42 osd.42 class hdd
device 43 osd.43 class hdd
device 44 osd.44 class hdd
device 45 osd.45 class hdd
device 46 osd.46 class hdd
device 47 osd.47 class hdd
device 48 osd.48 class hdd
device 49 osd.49 class hdd

# types
type 0 osd
type 1 host
type 2 chassis
type 3 rack
type 4 row
type 5 pdu
type 6 pod
type 7 room
type 8 datacenter
type 9 zone
type 10 region
type 11 root

# buckets
host server1 {
	id -3		# do not change unnecessarily
	id -16 class nvme		# do not change unnecessarily
	id -11 class hdd		# do not change unnecessarily
	# weight 31.765
	alg straw2
	hash 0	# rjenkins1
	item osd.18 weight 2.729
	item osd.19 weight 2.729
	item osd.22 weight 2.729
	item osd.23 weight 2.729
	item osd.24 weight 2.729
	item osd.34 weight 0.873
	item osd.16 weight 5.458
	item osd.20 weight 5.458
	item osd.38 weight 5.458
	item osd.1 weight 0.873
}
host server2 {
	id -5		# do not change unnecessarily
	id -17 class nvme		# do not change unnecessarily
	id -12 class hdd		# do not change unnecessarily
	# weight 31.765
	alg straw2
	hash 0	# rjenkins1
	item osd.4 weight 2.729
	item osd.5 weight 2.729
	item osd.6 weight 2.729
	item osd.8 weight 2.729
	item osd.9 weight 2.729
	item osd.32 weight 0.873
	item osd.7 weight 5.458
	item osd.10 weight 5.458
	item osd.31 weight 5.458
	item osd.3 weight 0.873
}
host server3 {
	id -7		# do not change unnecessarily
	id -18 class nvme		# do not change unnecessarily
	id -13 class hdd		# do not change unnecessarily
	# weight 31.765
	alg straw2
	hash 0	# rjenkins1
	item osd.11 weight 2.729
	item osd.12 weight 2.729
	item osd.14 weight 2.729
	item osd.17 weight 2.729
	item osd.2 weight 0.873
	item osd.15 weight 5.458
	item osd.36 weight 5.458
	item osd.13 weight 2.729
	item osd.37 weight 5.458
	item osd.33 weight 0.873
}
host server4 {
	id -9		# do not change unnecessarily
	id -19 class nvme		# do not change unnecessarily
	id -14 class hdd		# do not change unnecessarily
	# weight 31.765
	alg straw2
	hash 0	# rjenkins1
	item osd.25 weight 2.729
	item osd.27 weight 2.729
	item osd.26 weight 2.729
	item osd.28 weight 2.729
	item osd.30 weight 2.729
	item osd.35 weight 0.873
	item osd.0 weight 0.873
	item osd.21 weight 5.458
	item osd.29 weight 5.458
	item osd.39 weight 5.458
}
host server5 {
	id -2		# do not change unnecessarily
	id -4 class nvme		# do not change unnecessarily
	id -6 class hdd		# do not change unnecessarily
	# weight 31.766
	alg straw2
	hash 0	# rjenkins1
	item osd.40 weight 0.873
	item osd.41 weight 0.873
	item osd.42 weight 5.458
	item osd.43 weight 5.458
	item osd.44 weight 2.729
	item osd.45 weight 5.458
	item osd.46 weight 2.729
	item osd.47 weight 2.729
	item osd.48 weight 2.729
	item osd.49 weight 2.729
}
root default {
	id -1		# do not change unnecessarily
	id -20 class nvme		# do not change unnecessarily
	id -15 class hdd		# do not change unnecessarily
	# weight 158.826
	alg straw2
	hash 0	# rjenkins1
	item server1 weight 31.765
	item server2 weight 31.765
	item server3 weight 31.765
	item server4 weight 31.765
	item server5 weight 31.766
}

rule storage_metadata {
	id 2
	type replicated
	min_size 2
	max_size 3
	step take default class nvme
	step chooseleaf firstn 0 type host
	step emit
}
rule storage_data {
	id 3
	type erasure
	min_size 10
	max_size 11
	step set_chooseleaf_tries 5
	step set_choose_tries 100
	step take default class hdd
	step choose indep 4 type host
	step chooseleaf indep 3 type osd
	step emit
}
# end crush map
```

**EC Profile**
```sh
crush-device-class=hdd
crush-failure-domain=host
crush-root=default
jerasure-per-chunk-alignment=false
k=8
m=3
plugin=jerasure
technique=reed_sol_van
w=8
```

**Comment by TheJJ** [comment]
hai! yes this could be the reason, but i have to take a good look at these gory internals to be sure what's the correct fix. can you send me (e.g. via mail) your state dump as described in the readme? then i can reproduce the bug locally. thanks!

**Comment by BonzTM** [comment]
> hai! yes this could be the reason, but i have to take a good look at these gory internals to be sure what's the correct fix. can you send me (e.g. via mail) your state dump as described in the readme? then i can reproduce the bug locally. thanks!

Thanks!  Sent via email.

**Comment by TheJJ** [comment]
I think this is now fixed by b6d83ac3212aea2bc56300850d1a756f5efc24fe, and your suspicion about the cause was correct. so the fix for allowing less uses than the crush rule emits due to pool size should be fine :)
